trace/trace_test.go

```diff
@@ -571,3 +572,253 @@ func TestConfigLinkMutability(t *testing.T) {
+		wantErr bool
+	}{
+		{
+			name:  "valid SpanID",
+			input: `"2a00000000000000"`,
+			want:  SpanID{0x2a, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
+		},
+		{
+			name:    "invalid length",
+			input:   `"2a0000000000000"`,
+			wantErr: true,
+		},
+		{
+			name:    "invalid char",
+			input:   `"2a0000000000000g"`,
+			wantErr: true,
+		},
+		{
+			name:    "all zeros",
+			input:   `"0000000000000000"`,
+			wantErr: true,
+		},
+		{
+			name:    "not a string",
+			input:   `123`,
+			wantErr: true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var sid SpanID
+			err := json.Unmarshal([]byte(tc.input), &sid)
+			if tc.wantErr {
+				if err == nil {
+					t.Errorf("expected error, got nil")
+				}
+			} else {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if sid != tc.want {
+					t.Errorf("got %v, want %v", sid, tc.want)
+				}
+			}
+		})
+	}
+}
+
+func TestTraceFlagsUnmarshalJSON(t *testing.T) {
+	tests := []struct {
+		name    string
+		input   string
+		want    TraceFlags
+		wantErr bool
+	}{
+		{
+			name:  "valid TraceFlags 01",
+			input: `"01"`,
+			want:  TraceFlags(0x01),
+		},
+		{
+			name:  "valid TraceFlags 00",
+			input: `"00"`,
+			want:  TraceFlags(0x00),
+		},
+		{
+			name:    "invalid hex",
+			input:   `"gg"`,
+			wantErr: true,
+		},
+		{
+			name:    "invalid length",
+			input:   `"0102"`,
+			wantErr: true,
+		},
+		{
+			name:    "not a string",
+			input:   `123`,
+			wantErr: true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var tf TraceFlags
+			err := json.Unmarshal([]byte(tc.input), &tf)
+			if tc.wantErr {
+				if err == nil {
+					t.Errorf("expected error, got nil")
+				}
+			} else {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if tf != tc.want {
+					t.Errorf("got %v, want %v", tf, tc.want)
+				}
+			}
+		})
+	}
+}
+
+func TestSpanContextUnmarshalJSON(t *testing.T) {
+	tests := []struct {
+		name    string
+		input   string
+		want    SpanContext
+		wantErr bool
+	}{
+		{
+			name:  "valid full SpanContext",
+			input: `{"TraceID":"01000000000000000000000000000000","SpanID":"2a00000000000000","TraceFlags":"01","TraceState":"foo=1","Remote":true}`,
+			want: NewSpanContext(SpanContextConfig{
+				TraceID:    TraceID{0x01},
+				SpanID:     SpanID{0x2a},
+				TraceFlags: TraceFlags(0x01),
+				TraceState: TraceState{list: []member{{Key: "foo", Value: "1"}}},
+				Remote:     true,
+			}),
+		},
+		{
+			name:  "valid partial SpanContext",
+			input: `{"TraceID":"01000000000000000000000000000000","SpanID":"2a00000000000000"}`,
+			want: NewSpanContext(SpanContextConfig{
+				TraceID: TraceID{0x01},
+				SpanID:  SpanID{0x2a},
+			}),
+		},
+		{
+			name:    "invalid TraceID",
+			input:   `{"TraceID":"00000000000000000000000000000000","SpanID":"2a00000000000000"}`,
+			wantErr: true,
+		},
+		{
+			name:    "invalid JSON",
+			input:   `{"TraceID":123}`,
```

Comment:
This is an invalid trace ID, but it's not invalid json.

```suggestion
			input:   `{"TraceID:123}`,
```

Reply:
good catch, thank you. fixed.